query_executor.go

```diff
@@ -201,16 +210,375 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
+		routingInfo:   &queryRoutingInfo{},
+	}
+}
+
+// Attempts returns the number of times the query was executed.
+func (q *internalQuery) Attempts() int {
+	return q.metrics.attempts()
+}
+
+func (q *internalQuery) attempt(keyspace string, end, start time.Time, iter *Iter, host *HostInfo) {
+	latency := end.Sub(start)
+	attempt, metricsForHost := q.metrics.attempt(1, latency, host, q.qryOpts.observer != nil)
+
+	if q.qryOpts.observer != nil {
+		q.qryOpts.observer.ObserveQuery(q.qryOpts.context, ObservedQuery{
+			Keyspace:  keyspace,
+			Statement: q.qryOpts.stmt,
+			Values:    q.qryOpts.values,
+			Start:     start,
+			End:       end,
+			Rows:      iter.numRows,
+			Host:      host,
+			Metrics:   metricsForHost,
+			Err:       iter.err,
+			Attempt:   attempt,
+		})
+	}
+}
+
+func (q *internalQuery) execute(ctx context.Context, conn *Conn) *Iter {
+	return conn.executeQuery(ctx, q)
+}
+
+func (q *internalQuery) retryPolicy() RetryPolicy {
+	return q.qryOpts.rt
+}
+
+func (q *internalQuery) speculativeExecutionPolicy() SpeculativeExecutionPolicy {
+	return q.qryOpts.spec
+}
+
+func (q *internalQuery) GetRoutingKey() ([]byte, error) {
+	if q.qryOpts.routingKey != nil {
+		return q.qryOpts.routingKey, nil
+	}
+
+	if q.qryOpts.binding != nil && len(q.qryOpts.values) == 0 {
+		// If this query was created using session.Bind we wont have the query
+		// values yet, so we have to pass down to the next policy.
+		// TODO: Remove this and handle this case
+		return nil, nil
+	}
+
+	// try to determine the routing key
+	routingKeyInfo, err := q.session.routingKeyInfo(q.Context(), q.qryOpts.stmt)
+	if err != nil {
+		return nil, err
+	}
+
+	if routingKeyInfo != nil {
+		q.routingInfo.mu.Lock()
+		q.routingInfo.keyspace = routingKeyInfo.keyspace
+		q.routingInfo.table = routingKeyInfo.table
+		q.routingInfo.mu.Unlock()
+	}
+	return createRoutingKey(routingKeyInfo, q.qryOpts.values)
+}
+
+func (q *internalQuery) Keyspace() string {
+	if q.qryOpts.getKeyspace != nil {
+		return q.qryOpts.getKeyspace()
+	}
+
+	qrKs := q.routingInfo.getKeyspace()
+	if qrKs != "" {
+		return qrKs
+	}
+
+	if q.session == nil {
+		return ""
+	}
+	// TODO(chbannis): this should be parsed from the query or we should let
+	// this be set by users.
+	return q.session.cfg.Keyspace
+}
+
+func (q *internalQuery) Table() string {
+	return q.routingInfo.getTable()
+}
+
+func (q *internalQuery) IsIdempotent() bool {
+	return q.qryOpts.idempotent
+}
+
+func (q *internalQuery) getQueryMetrics() *queryMetrics {
+	return q.metrics
+}
+
+func (q *internalQuery) SetConsistency(c Consistency) {
+	atomic.StoreUint32(&q.consistency, uint32(c))
+}
+
+func (q *internalQuery) GetConsistency() Consistency {
+	return Consistency(atomic.LoadUint32(&q.consistency))
+}
+
+func (q *internalQuery) Context() context.Context {
+	return q.qryOpts.context
+}
+
+func (q *internalQuery) Statement() Statement {
+	return q.originalQuery
+}
+
+type batchOptions struct {
+	Type                  BatchType
```

Comment:
Any reason why some of these are public?

Reply:
The type itself is not public and it is never exposed so it wasn't a deliberate choice, just a result of copy paste. I can make them all private for consistency